core/typedef.go

```diff
@@ -72,6 +72,50 @@ func (fn *Function) WithArg(name string, typeDef *TypeDef, desc string, defaultV
 	return fn
 }
 
+func (fn *Function) IsSubtypeOf(otherFn *Function) bool {
+	if fn == nil || otherFn == nil {
+		return false
+	}
+
+	// check return type
+	if !fn.ReturnType.IsSubtypeOf(otherFn.ReturnType) {
+		return false
+	}
+
+	// check args
+	for i, otherFnArg := range otherFn.Args {
+		/* TODO: with more effort could probably relax and allow:
+		* arg names to not match (only types really matter in theory)
+		* mismatches in optional (provided defaults exist, etc.)
+		* fewer args in interface fn than object fn (as long as the ones that exist match)
+		 */
+
+		if i >= len(fn.Args) {
+			return false
+		}
+		fnArg := fn.Args[i]
+
+		if fnArg.Name != otherFnArg.Name {
+			return false
+		}
+
+		if fnArg.TypeDef.Optional != otherFnArg.TypeDef.Optional {
+			return false
+		}
```

Comment:
My thinking on defaults in terms of the permutations of "whether obj arg sets default" + "whether iface arg sets default" is:
1. Neither sets - doesn't matter
2. Both sets defaults of different values - just prefer the interface default since that's what it's "dressed up as" now anyways.
3. Only one or the other sets the default - just use that default

It's extremely subjective and a definite grey area, but that felt most intuitive to me.

The thing about defaults is that they don't actually have to matter for type checking. This is different than optional where it's actually important since an object arg that's non-optional can't be matched with an interface arg that's optional.